src/crypto/tls/handshake_server_tls13.go

```diff
@@ -370,15 +410,45 @@ func (hs *serverHandshakeStateTLS13) pickCertificate() error {
 		}
 		return err
 	}
+
 	hs.sigAlg, err = selectSignatureScheme(c.vers, certificate, hs.clientHello.supportedSignatureAlgorithms)
 	if err != nil {
 		// getCertificate returned a certificate that is unsupported or
 		// incompatible with the client's signature algorithms.
 		c.sendAlert(alertHandshakeFailure)
 		return err
 	}
+
 	hs.cert = certificate
 
+	if hs.clientHello.delegatedCredentialSupported && len(hs.clientHello.supportedSignatureAlgorithmsDC) > 0 && c.config.GetDelegatedCredential != nil {
+		dCred, priv, err := c.config.GetDelegatedCredential(clientHelloInfo(c, hs.clientHello), nil)
+		if err != nil {
+			c.sendAlert(alertInternalError)
+			return nil
+		}
+
+		if dCred != nil && priv != nil {
+			hs.cert.PrivateKey = priv
+			if dCred.raw == nil {
+				dCred.raw, err = dCred.marshal()
```

Comment:
in the delegated credential package it's possible to have an error where the signature is too long to be serialized, and this is only checked in marshal. I don't like this: a creation of a DC that doesn't fail should result in a DC that will work when put in TLS.

Reply:
I included a check at creation as well now.